File: leetcode_0083/java/leetcode_0083.java
import java.util.LinkedList;
import java.util.List;

public class leetcode_0083 {
    class ListNode<T>{
        T val;
        ListNode next;
        ListNode(){}
        ListNode(T val) {this.val = val;}
        ListNode(T val, ListNode next) {this.val = val;this.next = next;}
    }
    class Solution {
        public ListNode deleteDuplicates(ListNode head) {
            ListNode cur = head;
            if(cur == null)
                return null;
            while(cur.next != null){
                ListNode next = cur.next;
                if(cur.val == next.val){
                    cur.next = next.next;
                    next = null;
                }else{
                    cur = next;
                }
            }
            return head;
        }
        public  <T> void printVec(List<T> list){
            String res = "[";
            for(T ele : list){
                res += ele+",";
            }
            res = res.substring(0,res.length() - 1);
            res += "]";
            System.out.println(res);
        }

        public  <T> void printList(List<T> list){
            String res = "[";
            for(T ele : list){
                res += ele+",";
            }
            res = res.substring(0,res.length() - 1);
            res += "]";
            System.out.println(res);
        }

        public  <T> List<T> toList(leetcode_0083.ListNode head){
            List<T> list = new LinkedList<T>();
            leetcode_0083.ListNode cur = head;
            while(cur != null){
                list.add((T) cur.val);
                cur = cur.next;
            }
            return list;
        }
    }
    public static void main(String[] args) {
        Solution solution = new leetcode_0083().new Solution();
        ListNode head = new leetcode_0083().new ListNode(1,new leetcode_0083().new ListNode(1,new leetcode_0083().new ListNode(2,new leetcode_0083().new ListNode(3,new leetcode_0083().new ListNode(3)))));
        solution.deleteDuplicates(head);
        List<Integer> list = solution.toList(head);
        solution.printList(list);
        //result
        //[1,2,3]
    }
}

File: leetcode_0083/cpp/leetcode_0083.cpp
//
// Created by wangzy on 2022/3/17.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>

using namespace std;

struct ListNode{
    int val;
    ListNode *next;
    ListNode() : val(0),next(nullptr) {}
    ListNode(int x) : val(x),next(nullptr) {}
    ListNode(int x,ListNode *next):val(x),next(next) {}
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* cur = head;
        if(!cur)
            return nullptr;
        while(cur->next){
            ListNode* next = cur->next;
            if(cur->val == next->val){
                cur->next = next->next;
                next = nullptr;
            }
            else{
                cur = next;
            }
        }
        return head;
    }
};

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
void printList(std::list<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
std::list<T> toList(ListNode* head){
    std::list<T> l;
    ListNode* cur = head;
    while(cur){
        l.push_back(cur->val);
        cur = cur->next;
    }
    return l;
}

void delList(ListNode* head){
    ListNode* cur = head;
    while(cur && cur->next){
        ListNode* next = cur->next;
        cur = nullptr;
        cur = next;
    }
    cur = nullptr;
}

int main() {
    ListNode* l = new ListNode(1,new ListNode(1,new ListNode(2,new ListNode(3,new ListNode(3)))));
    ListNode* head = Solution().deleteDuplicates(l);
    list<int> li = toList<int>(head);
    printList(li);
    //输入：head = [1,1,2,3,3]
    //输出：[1,2,3]

}

